empathyco/x
# Set cache on queries preview results

[EMP-2974](https://searchbroker.atlassian.net/browse/EMP-2974) 


# Pull request template
 
We could face a memory leak if we use the Query Results Preview in the Empathize to show something as an Instant Search as we showed in the Sprint Review, in that case, we wouldn’t want to cache the results.

**CHANGES APPLIED**: 
* Add a cache prop and only cache the results accordingly (`saveCache` prop set to false by default, we will only catch the queries in` BrandRecommendations` to avoid requesting twice).
    * `saveCache` is set at true only with brand recommendations, because is the only feature where we want to save the queries. 

* Only trigger request if there are no cached results
    * When the `queryPreview` component is created we also check if the query is cached before triggering the request. If so, we emit `load` to render the queryPreview, otherwise, we emit `QueryPreviewRequestUpdated`. 

**PENDING TO RESOLVE:** --> [New task](https://searchbroker.atlassian.net/browse/EMP-3010): Note that in the current implementation, we are only saving queries. As a second step, we should also support filters, but then it should be the hash of query+filters+any other property that we might need or think about another workaround.

**PENDING TO RESOLVE:** --> Maybe we should deal with this in a separated PR
* Animation effect now it is linked to requests, DOM renders as soon as the response is received, by catching queries we lose that effect.
* Adding an animation component doesn’t solve the issue, it causes max call stack if used as a default in prop, or the lists are not rendered if passed as a computed prop. If this worked we could use the StaggeredFadeAndSlide when the caché is activated and ul when not. Before adding extraParams & filters, it worked only as a computed as follows: 
	```
        public get animation(): Vue | string | VueClass<Vue> {
  	  return StaggeredFadeAndSlide;
	} ```
* Adding a debounce when we emit load doesn’t solve the issue, then only one query preview is rendered and is kept in the loading state

## Motivation and context

Now, We are requesting the Query Preview search requests every time that they are rendered, so, every time that the user deletes the query.


- [x] Open issue. If applicable, link: [EMP-2867](https://searchbroker.atlassian.net/browse/EMP-2867)

## Type of change


## What is the destination branch of this PR?


## How has this been tested?


**TEST:**
* Open vue dev tools and check if brand recommendations are being saved in queryPreview state.
* Search for `battery` (has semantics) and check if brand recommendations still keep saving and if semantic queries are added to `queriesPreview` list.
* Clean the query in the search box and check the `queriesPreview` list. It should contain only the brand recommendations queries, semantic-queries have `saveCache` false, which means that these queries should be removed.

Tests performed according to [testing guidelines](./contributing/tests.md): 

## Checklist:



[EMP-2867]: https://searchbroker.atlassian.net/browse/EMP-2867?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[EMP-2974]: https://searchbroker.atlassian.net/browse/EMP-2974?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
